src/main/flight/alt_hold.c

```diff
@@ -80,47 +67,53 @@ float altitudePidCalculate(void)
     //   even though it may cause throttle oscillations while dropping quickly - the average D is what we need
     // - Prevent excessive iTerm growth when error is impossibly large for iTerm to resolve
 
-    const float altErrorCm = altHoldState.targetAltitudeCm - altHoldState.measuredAltitudeCm;
+    const pidCoefficient_t *pidCoefs =  getAltitudePidCoeffs();
+
+    const float altitudeErrorCm = altHoldState.targetAltitudeCm - getAltitudeCm();
 
     // P
-    const float pOut = simplePid.kp * altErrorCm;
+    const float pOut = pidCoefs->Kp * altitudeErrorCm;
 
     // I
     // input limit iTerm so that it doesn't grow fast with large errors
     // very important at the start if there are massive initial errors to prevent iTerm windup
 
-    // no iTerm change for error greater than 2m, otherwise it winds up badly
+    // much less iTerm change for errors greater than 2m, otherwise it winds up badly
     const float itermNormalRange = 200.0f; // 2m
-    const float itermRelax = (fabsf(altErrorCm) < itermNormalRange) ? 1.0f : 0.0f;
-    simplePid.integral += altErrorCm * taskIntervalSeconds * simplePid.ki * itermRelax;
+    const float itermRelax = (fabsf(altitudeErrorCm) < itermNormalRange) ? 1.0f : 0.1f;
+    pidIntegral += altitudeErrorCm * pidCoefs->Ki * itermRelax * taskIntervalSeconds;
     // arbitrary limit on iTerm, same as for gps_rescue, +/-20% of full throttle range
     // ** might not be needed with input limiting **
-    simplePid.integral = constrainf(simplePid.integral, -200.0f, 200.0f); 
-    const float iOut = simplePid.integral;
+    pidIntegral = constrainf(pidIntegral, -200.0f, 200.0f); 
+    const float iOut = pidIntegral;
 
     // D
     // boost D by 'increasing apparent velocity' when vertical velocity exceeds 5 m/s ( D of 75 on defaults)
-    // the velocity trigger is arbitrary at this point
     // usually we don't see fast ascend/descend rates if the altitude hold starts under stable conditions
     // this is important primarily to arrest pre-existing fast drops or climbs at the start;
 
-    float vel = altHoldState.smoothedVerticalVelocity;
+    float vel = getAltitudeDerivative(); // cm/s altitude derivative is always available
     const float kinkPoint = 500.0f; // velocity at which D should start to increase
     const float kinkPointAdjustment = kinkPoint * 2.0f; // Precompute constant
     const float sign = (vel > 0) ? 1.0f : -1.0f;
     if (fabsf(vel) > kinkPoint) {
         vel = vel * 3.0f - sign * kinkPointAdjustment;
     }
 
-    const float dOut = simplePid.kd * vel;
+    const float dOut = pidCoefs->Kd * -vel;
```

Comment:
Is the sign change intended? Looks like an error

Reply:
The sign change is correct.
Previously the calculation of smoothedVerticalVelocity was `previous - measured` so would be negative if climbing.  
But now the value from `getAltitudeDerivative()` is positive when climbing, so we need to invert the D value at some point.
We could perhaps subtract dOut but it seemed better to make all the PIDs add up.
Alternatively we could make the sign of `getAltitudeDerivative` the other way around, but I think positive for climbing is what it should be.